src/components/Firestore/DocumentEditor/index.tsx

```diff
@@ -85,272 +88,300 @@ export function supportsEditing(value: FirestoreAny): boolean {
  * areRootFielsMutable: can a root field be added/removed.
  */
 const DocumentEditor: React.FC<{
-  value: FirestoreMap;
+  value: FirestoreAny;
   onChange?: (value?: FirestoreMap) => void;
   areRootNamesMutable?: boolean;
   areRootFieldsMutable?: boolean;
   rtdb?: boolean;
 }> = ({
   value,
   onChange,
-  areRootNamesMutable = true,
-  areRootFieldsMutable = true,
-  rtdb: isRtdb,
+  areRootNamesMutable,
+  areRootFieldsMutable,
+  rtdb = false,
 }) => {
+  const initialState = normalize(value);
+  const [store, dispatch] = React.useReducer(storeReducer, initialState);
   const methods = useForm({ mode: 'onChange' });
+  const api = useOptionalApi();
+
+  const denormalizedStore = React.useMemo(() => denormalize(store, api), [
+    store,
+    api,
+  ]);
+
+  const errorCount = React.useMemo(() => Object.keys(methods.errors).length, [
+    methods,
+  ]);
+
+  useEffect(() => {
+    if (errorCount === 0) {
+      onChange && onChange(denormalizedStore as FirestoreMap);
+    } else {
+      onChange && onChange(undefined);
+    }
+  }, [denormalizedStore, errorCount, onChange]);
 
   return (
     <FormContext {...methods}>
-      <DocumentProvider value={value}>
-        <RootEditor
-          onChange={onChange}
-          areRootNamesMutable={areRootNamesMutable}
-          areRootFieldsMutable={areRootFieldsMutable}
-          isRtdb={isRtdb}
-        />
-      </DocumentProvider>
+      <DocumentStore store={store} dispatch={dispatch}>
+        <div className="DocumentEditor">
+          {
+            <>
+              {store.id !== undefined && (
+                <FieldEditor
+                  id={store.id}
+                  isRtdb={rtdb}
+                  areNamesMutable={areRootNamesMutable}
+                  areFieldsMutable={areRootFieldsMutable}
+                />
+              )}
+            </>
+          }
+        </div>
+      </DocumentStore>
     </FormContext>
   );
 };
 
-function useErrorCount() {
-  const { errors } = useFormContext();
-  return Object.keys(errors).length;
-}
-
 /**
- * Special representation of a Document Root, where we don't want to show
- * the implicit top-level map.
+ * Field with call-to-actions for editing as well as rendering applicable child-nodes
  */
-const RootEditor: React.FC<{
-  onChange?: (value?: FirestoreMap) => void;
-  areRootNamesMutable: boolean;
-  areRootFieldsMutable: boolean;
-  isRtdb?: boolean;
-}> = ({ onChange, areRootNamesMutable, areRootFieldsMutable, isRtdb }) => {
-  const rootFields = useRootFields();
-  const dispatch = useDocumentDispatch()!;
-  const sdkMap = useSdkMap();
-  const errorCount = useErrorCount();
-
-  useEffect(() => {
-    async function validate() {
-      if (onChange) {
-        if (errorCount === 0) {
-          onChange(sdkMap);
-        } else {
-          onChange(undefined);
-        }
-      }
-    }
+const FieldEditor: React.FC<{
+  id: number;
```

Comment:
Can we make this id `uuid` or `guid` throughout, so that it's clear it's **not** the key of the field?

Reply:
Good idea, replaced with UUID